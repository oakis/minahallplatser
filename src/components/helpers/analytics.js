import analytics from '@react-native-firebase/analytics';

const convertSpaces = str => str.replace(/ /g, '_');

export const track = (title, props) => {
  if (props) {
    window.log(`Firebase analytics track with props: ${title}`, props);
    analytics().logEvent(convertSpaces(title), props);
  } else {
    window.log(`Firebase analytics track: ${title}`);
    analytics().logEvent(convertSpaces(title), {});
  }
};
