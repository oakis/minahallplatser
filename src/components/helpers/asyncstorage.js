import { AsyncStorage } from 'react-native';

export const generateUid = () => {
    return `${s4()}${s4()}-${s4()}-${s4()}-${s4()}-${s4()}${s4()}${s4()}`;
};

function s4() {
    return Math.floor((1 + Math.random()) * 0x10000).toString(16).substring(1);
}

/**
 * @param {string} item 
 * @returns {Promise} Loaded storage
 */
export const getStorage = (item) => {
    return new Promise((resolve, reject) => {
        AsyncStorage.getItem(item)
        .then((json) => {
            const data = json === null ? {} : JSON.parse(json);
            window.log(`getStorage() ${item}: OK`, data);
            resolve(data);
        })
        .catch((e) => {
            window.log(`getStorage() ${item}: FAIL`, e);
            reject(e);
        });
    });
};

/**
 * @param {string} item 
 * @param {JSON} json 
 * @returns {Promise} -
 */
export const setStorage = (item, json) => {
    return new Promise((resolve, reject) => {
        AsyncStorage.setItem(item, JSON.stringify(json))
        .then(() => {
            window.log(`setStorage() ${item}: OK`);
            resolve();
        })
        .catch((e) => {
            window.log(`setStorage() ${item}: FAIL`, e);
            reject(e);
        });
    });
};
