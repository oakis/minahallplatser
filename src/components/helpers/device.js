import { Platform } from 'react-native';

export const isAndroid = () => {
    return Platform.OS === 'android';
};

export const isIOS = () => {
    return Platform.OS === 'ios';
};
