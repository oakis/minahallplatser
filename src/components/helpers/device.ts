import {Platform} from 'react-native';
import {
  getModel,
  getSystemVersion,
  getVersion,
  getUniqueId,
} from 'react-native-device-info';

export const isAndroid = (): boolean => Platform.OS === 'android';

export const isIOS = (): boolean => Platform.OS === 'ios';

export const getDeviceModel = (): string => getModel();

export const getOsVersion = (): string => getSystemVersion();

export const getAppVersion = (): string => getVersion();

export const deviceId = (): string => getUniqueId();
