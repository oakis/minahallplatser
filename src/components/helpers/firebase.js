import fetch from 'react-native-cancelable-fetch';
import { firebaseFunctionsUrl } from '../../Server';
import { handleJsonFetch } from './';

export const updateDeparturesCount = (count) => {
    if (__DEV__) {
        return;
    }
    const url = `${firebaseFunctionsUrl}/addDeparturesCount?count=${count}`;
    fetch(url, {}, 'updateDeparturesCount')
    .finally(handleJsonFetch)
    .then(({ message }) => {
        window.log(`updateDeparturesCount(): OK (${message})`);
    })
    .catch((err) => {
        window.log(`updateDeparturesCount(): FAILED' ${message})`);
        new Error(err);
    });
};

export const getDeparturesCount = () => {
    const url = `${firebaseFunctionsUrl}/getDeparturesCount`;
    return fetch(url, {}, 'getDeparturesCount')
    .finally(handleJsonFetch)
    .then(({ departuresCount }) => {
        window.log('getDeparturesCount(): OK', departuresCount);
        return departuresCount;
    })
    .catch((err) => {
        window.log('getDeparturesCount(): FAILED', err);
        new Error(err);
    });
};

export const updateStopsCount = () => {
    if (__DEV__) {
        return;
    }
    const url = `${firebaseFunctionsUrl}/addStopsCount`;
    fetch(url, {}, 'updateStopsCount')
    .finally(handleJsonFetch)
    .then(({ message }) => {
        window.log('updateStopsCount(): OK', message);
    })
    .catch((err) => {
        window.log('updateStopsCount(): FAILED', message);
        new Error(err);
    });
};

export const getStopsCount = () => {
    const url = `${firebaseFunctionsUrl}/getStopsCount`;
    return fetch(url, {}, 'getStopsCount')
    .finally(handleJsonFetch)
    .then(({ stopsCount }) => {
        window.log('getStopsCount(): OK', stopsCount);
        return stopsCount;
    })
    .catch((err) => {
        window.log('getStopsCount(): FAILED', err);
        new Error(err);
    });
};

export const incrementStopsOpened = (user, stopId) => {
    const url = `${firebaseFunctionsUrl}/incrementStopsOpen?user=${user}&stopId=${stopId}`;
    fetch(url, {}, 'incrementStopsOpened')
    .finally(handleJsonFetch)
    .then(({ message }) => {
        window.log('incrementStopsOpened(): OK', message);
    })
    .catch((err) => {
        window.log('incrementStopsOpened(): FAILED', err);
        new Error(err);
    });
}