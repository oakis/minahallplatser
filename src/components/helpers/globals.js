export const globals = {
    shouldExitApp: false,
    didLogout: false,
    isLoggingIn: false
};
