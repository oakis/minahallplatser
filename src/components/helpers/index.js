export * from './device';
export * from './message';
export * from './network';
export * from './token';
export * from './firebase';
export * from './analytics';
export * from './globals';
export * from './image';

export const isObject = (obj) => {
    return Object.prototype.toString.call(obj) === '[object Object]';
};
