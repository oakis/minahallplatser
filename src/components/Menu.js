import React, { Component } from 'react';
import { View, AsyncStorage, ImageBackground, ScrollView } from 'react-native';
import { connect } from 'react-redux';
import firebase from 'react-native-firebase';
import { Actions } from 'react-native-router-flux';
import { ListItem, ListHeading } from './common';
import { RESET_ALL } from '../actions/types';
import { setSetting } from '../actions';
import { store } from '../App';
import { colors, metrics } from './style';
import { track, globals } from './helpers';
import { Feedback } from './modals';

class Menu extends Component {

    state = {
        user: firebase.auth().currentUser || { email: null },
        feedbackVisible: false,
    }

    logout() {
        globals.didLogout = true;
        globals.isLoggingIn = false;
        firebase.auth().signOut().then(() => {
            AsyncStorage.clear().then(() => {
                store.dispatch({ type: RESET_ALL });
                track('Logout', { Success: true });
            });
        }, (error) => {
            track('Logout', { Success: false });
            window.log('Sign Out Error', error);
        });
    }

    openFeedback() {
        this.setState({ feedbackVisible: true });
    }

    closeFeedback() {
        this.setState({ feedbackVisible: false });
    }

	render() {
		return (
			<View style={{ flexDirection: 'column', backgroundColor: colors.background, flex: 1 }}>
                <Feedback
                    visible={this.state.feedbackVisible}
                    close={() => this.closeFeedback()}
                />
                <ScrollView>
                    <ImageBackground source={{ uri: 'https://www.w3schools.com/css/img_fjords.jpg' }} style={{ width: 225, height: 120 }}>
                        <View
                            style={{
                                backgroundColor: 'rgba(255,255,255,0.7)',
                                flex: 1
                            }}
                        />
                    </ImageBackground>

                    <ListHeading text="Konto" />

                    {this.state.user && !this.state.user.isAnonymous ?
                        <ListItem
                            text="Profil"
                            icon="ios-person"
                            iconVisible
                            pressItem={() => {
                                Actions.profile();
                            }}
                        />
                    : null}

                    <ListItem
                        text="Inställningar"
                        icon="ios-settings"
                        iconVisible
                        pressItem={() => {
                            Actions.settings();
                        }}
                    />

                    <View style={{ marginBottom: metrics.margin.md }} />

                    <ListHeading text="Åtgärder" />

                    <ListItem
                        text="Lämna feedback"
                        icon="ios-mail-outline"
                        iconVisible
                        pressItem={() => {
                            track('Feedback Open');
                            this.openFeedback();
                        }}
                        style={{ marginTop: metrics.margin.md }} // Första ListItem ska ha en marginTop för att få ett jämnt mellanrum mellan ListItem's
                    />

                    {this.state.user && this.state.user.isAnonymous ?
                    <View>
                        <ListItem
                            text="Logga in"
                            icon="ios-log-in-outline"
                            iconVisible
                            pressItem={() => {
                                Actions.login();
                            }}
                        />
                        <ListItem
                            text="Registrera"
                            icon="ios-log-in-outline"
                            iconVisible
                            pressItem={() => {
                                Actions.register();
                            }}
                        />
                    </View>
                    :
                    <ListItem
                        text='Logga ut'
                        icon='ios-exit-outline'
                        iconVisible
                        pressItem={() => {
                            this.logout();
                        }}
                    />}
                </ScrollView>
            </View>
		);
	}

}

const mapStateToProps = state => {
    const { timeFormat, favoriteOrder, allowedGPS } = state.settings;
	return { favoriteOrder, timeFormat, allowedGPS };
};

export default connect(mapStateToProps, { setSetting })(Menu);
