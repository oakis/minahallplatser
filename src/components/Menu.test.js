import React from 'react';
import { shallow } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import firebase from 'react-native-firebase';
import { AsyncStorage } from 'react-native';
import { stub } from 'sinon';
import { globals, track } from './helpers';
import { store } from '../App';
import Menu from './Menu';


const mockStore = configureMockStore([thunk]);

const initialState = {
    settings: {
        timeFormat: 'minutes',
        favoriteOrder: 'nothing',
        allowedGPS: true,
    }
};

const oldAuth = firebase.auth();

const setSetting = stub();
beforeEach(() => {
    setSetting.reset();
});

it('should match snapshot', () => {
    const wrapper = shallow(<Menu store={mockStore(initialState)} />).dive();
    expect(wrapper).toMatchSnapshot();
});

it('should match snapshot if anonymous', () => {
    firebase.auth = () => ({ currentUser: { isAnonymous: true } });
    const wrapper = shallow(<Menu store={mockStore(initialState)} />).dive();
    expect(wrapper).toMatchSnapshot();
});

describe('logout', () => {
    let wrapper;
    store.dispatch = jest.fn();    
    beforeEach(async () => {
        firebase.auth = () => oldAuth;
        firebase.auth().signOut.reset();
        firebase.auth().signOut.resolves({});
        AsyncStorage.clear.mockClear();
        wrapper = shallow(<Menu store={mockStore(initialState)} />).dive();
        await wrapper.instance().logout();
    });
    it('should set globals correctly', () => {
        expect(globals.didLogout).toBe(true);
        expect(globals.isLoggingIn).toBe(false);
    });

    it('should sign out from firebase', () => {
        expect(firebase.auth().signOut.callCount).toBe(1);
    });

    it('should clear AsyncStorage', () => {
        expect(AsyncStorage.clear).toHaveBeenCalledTimes(1);
    });

    it('should be tracked on success', () => {
        expect(track).toBeCalledWith('Logout', { Success: true });
    });

    it('should reset store', () => {
        expect(store.dispatch).toBeCalledWith({ type: 'RESET_ALL' });
    });
});

it('openFeedback should set feedbackVisible to true', () => {
    const wrapper = shallow(<Menu store={mockStore(initialState)} />).dive();
    wrapper.instance().openFeedback();
    expect(wrapper.state().feedbackVisible).toBe(true);
});

it('closeFeedback should set feedbackVisible to false', () => {
    const wrapper = shallow(<Menu store={mockStore(initialState)} />).dive();
    wrapper.setState({ feedbackVisible: true });
    wrapper.instance().closeFeedback();
    expect(wrapper.state().feedbackVisible).toBe(false);
});
