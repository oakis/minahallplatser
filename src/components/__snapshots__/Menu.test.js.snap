// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#EAE9EF",
      "flex": 1,
      "flexDirection": "column",
    }
  }
>
  <Feedback
    close={[Function]}
    visible={false}
  />
  <ScrollView>
    <ImageBackground
      source={
        Object {
          "uri": "https://www.w3schools.com/css/img_fjords.jpg",
        }
      }
      style={
        Object {
          "height": 120,
          "width": 225,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "rgba(255,255,255,0.7)",
            "flex": 1,
          }
        }
      />
    </ImageBackground>
    <View
      style={
        Object {
          "height": 60,
        }
      }
    >
      <DefaultFont
        style={
          Object {
            "fontSize": 12,
            "fontWeight": "bold",
            "marginBottom": 0,
            "paddingBottom": 0,
            "paddingLeft": 10,
            "paddingTop": 10,
          }
        }
      >
        SORTERA FAVORITER
      </DefaultFont>
      <Picker
        itemStyle={
          Object {
            "fontSize": 16,
            "height": 90,
            "marginTop": -10,
          }
        }
        onValueChange={[Function]}
        selectedValue="nothing"
        style={
          Array [
            Object {},
          ]
        }
      >
        <Component
          label="Ingen sortering"
          value="nothing"
        />
        <Component
          label="Mina mest använda"
          value="opened"
        />
        <Component
          label="Efter bokstav"
          value="busStop"
        />
      </Picker>
    </View>
    <ListItemSeparator />
    <View
      style={
        Object {
          "height": 60,
        }
      }
    >
      <DefaultFont
        style={
          Object {
            "fontSize": 12,
            "fontWeight": "bold",
            "marginBottom": 0,
            "paddingBottom": 0,
            "paddingLeft": 10,
            "paddingTop": 10,
          }
        }
      >
        TIDSFORMAT
      </DefaultFont>
      <Picker
        itemStyle={
          Object {
            "fontSize": 16,
            "height": 90,
            "marginTop": -10,
          }
        }
        onValueChange={[Function]}
        selectedValue="minutes"
        style={
          Array [
            Object {},
          ]
        }
      >
        <Component
          label="Minuter"
          value="minutes"
        />
        <Component
          label="Klockslag"
          value="clock"
        />
      </Picker>
    </View>
    <ListItemSeparator />
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    />
    <ListItem
      icon="email"
      iconVisible={true}
      pressItem={[Function]}
      style={
        Object {
          "marginTop": 10,
        }
      }
      text="Lämna feedback"
    />
  </ScrollView>
</View>
`;
